gep: standardizing behavior for invalid BackendTLSPolicy (#3909)
kubernetes-sigs/gateway-api: geps/gep-1897/index.md

named object references, each containing a single cert. We originally proposed to follow the convention established by the
[CertificateRefs field on Gateway](https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340)
, but the CertificateRef requires both a tls.key and tls.crt and a certificate reference only requires the tls.crt.

> If any of the CACertificateRefs cannot be resolved (e.g., the referenced resource does not exist) or is misconfigured (e.g., ConfigMap does not contain a key named `ca.crt`), the `ResolvedRefs` status condition MUST be set to `False` with `Reason: InvalidCACertificateRef` and a connection using that CACertificateRef MUST fail, and the client MUST receive an HTTP 5xx error response.

Author: Added that implementations should indicate invalid references in the message, as also suggested by @youngnick.

> Any further validation of the certificate content (i.e., checking expiry or enforcing specific formats) is implementation-specific.

Reviewer: ```suggestion
Implemenmtations MAY perform further validation of the certificate content (i.e., checking expiry or enforcing specific formats). If they do, they MUST ensure that the `ResolvedRefs` Condition is `false`, with a `message` indicating which references are invalid, and why, and use a domain-prefixed implementation-specific `reason`, like `implementation.io/ExpiredCertificate` or similar).
```

Author: The `Reason` must follow the pattern `^[A-Za-z]([A-Za-z0-9_,:]*[A-Za-z0-9_])?$`. Therefore, domain-prefixed reasons are not supported.

named object references, each containing a single cert. We originally proposed to follow the convention established by the
[CertificateRefs field on Gateway](https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340)
, but the CertificateRef requires both a tls.key and tls.crt and a certificate reference only requires the tls.crt.
If any of the CACertificateRefs cannot be resolved (e.g., the referenced resource does not exist) or is misconfigured (e.g., ConfigMap does not contain a key named `ca.crt`), the `ResolvedRefs` status condition MUST be set to `False` with `Reason: InvalidCACertificateRef`. Connections using that CACertificateRef MUST fail, and the client MUST receive an HTTP 5xx error response.
References to objects with an unsupported Group and Kind are not valid, and MUST be rejected by the implementation with the `ResolvedRefs` status condition set to `False` and `Reason: UnsupportedFeature`.

> Implementations MAY perform further validation of the certificate content (i.e., checking expiry or enforcing specific formats). If they do, they MUST ensure that the `ResolvedRefs` Condition is `False` and use an implementation-specific `Reason`, like `ExpiredCertificate` or similar.

Reviewer: I would prefer that implementation-specific Condition things use a domain-prefixed name, like `implementation.io/ExpiredCertificate`, otherwise if we ever do end up adding this to Gateway API itself, the "no domain" namespace will have certain words used by implementations. That's what the domain-prefixing is for, to prevent use of words we might want in the core API later.

This isn't a a hard blocker, just a nice to have, because this is only a Reason rather than a Type.

named object references, each containing a single cert. We originally proposed to follow the convention established by the
[CertificateRefs field on Gateway](https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340)
, but the CertificateRef requires both a tls.key and tls.crt and a certificate reference only requires the tls.crt.

> If any of the CACertificateRefs cannot be resolved or is misconfigured, the BackendTLSPolicy is considered invalid.

Author: According to the Spec (https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha3/backendtlspolicy_types.go#L121C1-L123C61)  the behavior for attaching multiple CACertificateRefs is implementation specific:
> A single CACertificateRef to a Kubernetes ConfigMap kind has "Core" support. Implementations MAY choose to support attaching multiple certificates to a backend, but this behavior is implementation-specific.

```suggestion
If all of the CACertificateRefs cannot be resolved or are misconfigured, the BackendTLSPolicy is considered invalid.
If an implementation chooses to support attaching multiple certificates, the behavior is implementation-specific.
```

AFAIK, only `envoy-gateway` distinguishes between cases where all or any CertificateRefs are unresolvable.

\cc @kl52752 @robscott

Reviewer 2: I have concerns about allowing customers to attach partially valid BackendTLSPolicy configurations by default. If a customer omits some references, it could lead to intermittent connectivity issues because only a portion of the configurations would be used to communicate with the backend. This could result in a confusing and unreliable experience for the customer.

My proposition is to:

- If any of the CACertificateRefs cannot be resolved or is misconfigured, the BackendTLSPolicy is considered invalid. 
- If a newly added BackendTLSPolicy is invalid implementation should reject this object. 
- If an existing BackendTLSPolicy is updated to an invalid object it is implementation specific whether this object should be rejected. However, the implementation must clearly signal to the user the current configuration and why the object is invalid.

> If the CertificateRef cannot be resolved or does not include a certificate (tls.crt), the BackendTLSPolicy is considered invalid.

Reviewer 3: The type of CACertificateRefs is []LocalObjectReference, so I don't think we can insist on this level of validation from the implementation.  How about:
```suggestion
If any of the CACertificateRefs cannot be resolved or is misconfigured, the BackendTLSPolicy is considered invalid.
```

Author: I'm fine with either approach, as long as the behavior is clearly defined.

However, I would prefer that if any CertificateRef cannot be resolved, the entire BackendTLSPolicy should be considered invalid. This provides a clear and consistent model, rather than introducing ambiguity with partially invalid policies that behave differently from completely invalid ones.

named object references, each containing a single cert. We originally proposed to follow the convention established by the
[CertificateRefs field on Gateway](https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340)
, but the CertificateRef requires both a tls.key and tls.crt and a certificate reference only requires the tls.crt.

> If any of the CACertificateRefs cannot be resolved (e.g., the referenced resource does not exist) or is misconfigured (e.g., ConfigMap does not contain a key named `ca.crt`), the BackendTLSPolicy is considered invalid.

Reviewer 3: This point should be clarified to include the concept that one invalid CACertificateRef doesn't make the entire BTP Invalid:true or Accepted:false.  However, it must block the TLS connection if an invalid CACertificateRef is attempted to be used.

```suggestion
If any of the CACertificateRefs cannot be resolved (e.g., the referenced resource does not exist) or is misconfigured (e.g., ConfigMap does not contain a key named `ca.crt`), a connection using that CACertificateRef must fail.
```

If ClientCertificateRefs is unspecified, then WellKnownCACertificates must be set to "System" for a valid configuration.
If WellKnownCACertificates is unspecified, then CACertificateRefs must be specified with at least one entry for a valid configuration.
If WellKnownCACertificates is set to "System" and there are no system trusted certificates or the implementation doesn't define system

> trusted certificates, the BackendTLSPolicy is considered invalid.

Reviewer 3: Again, I don't think we can insist on this level of validation except at runtime.  There may not be a system trusted certificate when the policy is created, but we shouldn't mark it invalid unless the connection fails because of that.  Alternatively, there may be a system trusted certificate when it is created, but it expires at some time in the future and makes the connection fail.  

I don't think we want the implementations to set up watches on these certificates.

Reviewer 3: Fair question, but how is the implementation supposed to determine whether WellKnownCACertificates are supported at the time when it sets the status for the BackendTLSPolicy?

Author: I think we might be talking past each other a bit. What I was referring to is already captured in the API specification:
```
// If an implementation does not support the WellKnownCACertificates field,
// or the provided value is unsupported, the Policy's Status Conditions MUST be
// updated to include an Accepted: False Condition with Reason: Invalid.
```
https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha3/backendtlspolicy_types.go#L140

> For an invalid BackendTLSPolicy, implementations MUST NOT fall back to unencrypted (plaintext) connections. 

Reviewer 2: I think that "Invalid" isn't the right term here because BTP spec is validated by CEL, making an invalid BTP impossible.

If this refers to a hostname or certificate mismatch with backend configuration, please state that explicitly.
Besides, this case is already covered with implementation specific way. 
.

```
On the question of how to signal that there was a failure in the certificate validation, this is left up to the implementation to return a response error that is appropriate, such as one of the HTTP error codes: 400 (Bad Request), 401 (Unauthorized), 403 (Forbidden), or other signal that makes the failure sufficiently clear to the requester without revealing too much about the transaction, based on established security requirements.
```

Author: As we discussed off-channel, I had two cases in mind where I would consider a BackendTLSPolicy to be invalid, apart from the standard CEL validation or runtime errors:
> If the CertificateRef cannot be resolved or does not include a certificate (tls.crt), the BackendTLSPolicy is considered invalid.

> If WellKnownCACertificates is set to "System" and there are no system trusted certificates or the implementation doesn't define system
trusted certificates, the BackendTLSPolicy is considered invalid.

My main concern for raising this is to understand, how would the BackendTLSPolicy signal that the referenced resource in CertificateRef does not exist or cannot be resolved?

Reviewer 2: We can created new conditions similar to "ResolvedRefs" condition which is dedicated for listeners?

Instead, the corresponding TLS connection MUST fail, and the client MUST receive an HTTP error response.

> Additionally, the `Accepted` status condition of the BackendTLSPolicy MUST be set to `False` with the reason `Invalid`.

Author: Looking at it again, it might make sense to introduce the `ResolvedRefs` condition for policies as well. However, I’m not sure whether it fits within the current schedule for graduating the BackendTLSPolicy.

Reviewer 2: It might be difficult to implement due to the distinct processes of policy application and certificate validation against the backend. It's a valid scenario that the BTP was accepted, and configuration properly propagated by the controller, but connectivity is broken due to certificate misconfiguration. All information needed to debug this issue should be passed by inspecting the BTP spec against the Service to which the Policy was applied.

Reviewer 4: Agree with @kl52752 that ResolvedRefs would be difficult to set for anything that was tied to dataplane/connectivity. On the other hand, ResolvedRefs seems like a useful concept on BackendTLSPolicy for the case that a CACertRef is invalid.

Reviewer 2: this topic is discussed in multiple threads :) https://github.com/kubernetes-sigs/gateway-api/pull/3909/files#r2230306410

Reviewer: To summarize: I think that ResolvedRefs should be included and used for certificates that do not exist or are not valid (where "not valid" means "does not contain certain keys", not "unwrap the certificate and check certificate properties").

Decoding the certificate should never be _required_ for an implementation to be conformant. Implementations MAY unwrap if they wish and do additional error handling, but those errors should be in addition to the included error handling.

Reviewer: Agreed that adding some explicit examples of what makes a reference invalid is fine - _but_ the second item - "ref exist but does not contain the right keys" requires decoding the Secret, so it can only be a MAY.

between the gateway and backend pod. References to a resource in a different namespace are invalid.
If ClientCertificateRefs is unspecified, then WellKnownCACertificates must be set to "System" for a valid configuration.
If WellKnownCACertificates is unspecified, then CACertificateRefs must be specified with at least one entry for a valid configuration.

> If an implementation does not support the WellKnownCACertificates, or the provided value is unsupported,the BackendTLSPolicy is considered invalid.

Reviewer: ```suggestion
If an implementation does not support the WellKnownCACertificates, or the provided value is unsupported,the BackendTLSPolicy is considered invalid, and the implementation MUST set the `Accepted` Condition to `false`, with a reason of `UnsupportedFeature` and a message explaining this.
```
